pyomo/contrib/parmest/parmest.py

```diff
@@ -449,65 +393,46 @@ def _Q_opt(self, ThetaVals=None, solver="ef_ipopt",
-        construct the tree just once and reuse it, then remember to
-        remove thetavals from it when none is desired.
         """
-        
-        assert(solver != "k_aug" or ThetaVals == None)
-        # Create a tree with dummy scenarios (callback will supply when needed).
-        # Which names to use (i.e., numbers) depends on if it is for bootstrap.
+        if (solver == "k_aug"):
+            raise RuntimeError("k_aug no longer supported.")
+
         # (Bootstrap scenarios will use indirection through the bootlist)
         if bootlist is None:
-            tree_model = _treemaker(self._numbers_list)
+            scen_names = ["Scenario{}".format(i) for i in self._numbers_list]
         else:
-            tree_model = _treemaker(range(len(self._numbers_list)))
-        stage1 = tree_model.Stages[1]
-        stage2 = tree_model.Stages[2]
-        tree_model.StageVariables[stage1] = self.theta_names
-        tree_model.StageVariables[stage2] = []
-        tree_model.StageCost[stage1] = "FirstStageCost"
-        tree_model.StageCost[stage2] = "SecondStageCost"
-
-        # Now attach things to the tree_model to pass them to the callback
-        tree_model.CallbackModule = None
-        tree_model.CallbackFunction = self._instance_creation_callback
+            scen_names = ["Scenario{}".format(i)\
+                         for i in range(len(self._numbers_list))]
+
+        # tree_model.CallbackModule = None
+        outer_cb_data = dict()
+        outer_cb_data["callback"] = self._instance_creation_callback
         if ThetaVals is not None:
-            tree_model.ThetaVals = ThetaVals
+            outer_cb_data["ThetaVals"] = ThetaVals
         if bootlist is not None:
-            tree_model.BootList = bootlist
-        tree_model.cb_data = self.callback_data  # None is OK
-
-        try:
-            # For structured models, it is important that we use the
-            # updated version of the CUID representation.  PySP (for
-            # backwards compatibility reasons, and so a ton of tests
-            # don't have to be updated) still defaults to the old
-            # representation.
-            _cuidver = scenariotree.tree_structure.CUID_repr_version
-            scenariotree.tree_structure.CUID_repr_version = 2
-            stsolver = st.StochSolver(
-                fsfile = "pyomo.contrib.parmest.parmest",
-                fsfct = "_pysp_instance_creation_callback",
-                tree_model = tree_model
-            )
-        finally:
-            scenariotree.tree_structure.CUID_repr_version = _cuidver
-                
+            outer_cb_data["BootList"] = bootlist
+        outer_cb_data["cb_data"] = self.callback_data  # None is OK
+        outer_cb_data["theta_names"] = self.theta_names
+
+        options = {"solver": "ipopt"}
+        scenario_creator_options = {"cb_data": outer_cb_data}
+        if use_mpisppy:
+            ef = sputils.create_EF(scen_names,
```

Comment:
Similar question here as above, can we just explicitly import `create_EF` and avoid this if-tree?

Reply:
I think we should leave it this way to help future maintainers.